stateresolutionv2.go

```diff
@@ -0,0 +1,459 @@
+		resolvedMembers:           make(map[string]*Event),
+	}
+
+	// Separate out power level events from the rest of the events. This is
+	// necessary because we perform topological ordering of the power events
+	// separately, and then the mainline ordering of all other events depends
+	// on that power level ordering.
+	for _, p := range conflicted {
+		if p.Type() == MRoomPowerLevels {
+			conflictedPowerLevels = append(conflictedPowerLevels, p)
+		} else {
+			conflictedOthers = append(conflictedOthers, p)
+		}
+	}
+
+	// Start with the unconflicted events by ordering them topologically and then
+	// authing them. The successfully authed events will form the initial partial
+	// state.
+	unconflicted = r.reverseTopologicalOrdering(unconflicted)
+	r.authAndApplyEvents(unconflicted)
+
+	// Then order the conflicted power level events topologically and then also
+	// auth those too. The successfully authed events will be layered on top of
+	// the partial state.
+	conflictedPowerLevels = r.reverseTopologicalOrdering(conflictedPowerLevels)
+	r.authAndApplyEvents(conflictedPowerLevels)
+
+	// Then generate the mainline of power level events, order the remaining state
+	// events based on the mainline ordering and auth those too. The successfully
+	// authed events are also layered on top of the partial state.
+	r.powerLevelMainline = r.createPowerLevelMainline()
+	r.authAndApplyEvents(r.mainlineOrdering(conflictedOthers))
+
+	// Finally we will reapply the original set of unconflicted events onto the
+	// partial state, just in case any of these were overwritten by pulling in
+	// auth events in the previous two steps, and that gives us our final resolved
+	// state.
+	r.authAndApplyEvents(unconflicted)
+
+	// Now that we have our final state, populate the result array with the
+	// resolved state and return it.
+	if r.resolvedCreate != nil {
+		r.result = append(r.result, *r.resolvedCreate)
+	}
+	if r.resolvedJoinRules != nil {
+		r.result = append(r.result, *r.resolvedJoinRules)
+	}
+	if r.resolvedPowerLevels != nil {
+		r.result = append(r.result, *r.resolvedPowerLevels)
+	}
+	for _, member := range r.resolvedMembers {
+		r.result = append(r.result, *member)
+	}
+	for _, invite := range r.resolvedThirdPartyInvites {
+		r.result = append(r.result, *invite)
+	}
+	return r.result
+}
+
+// createPowerLevelMainline generates the mainline of power level events,
+// starting at the currently resolved power level event from the topological
+// ordering and working our way back to the room creation. Note that we populate
+// the result here in reverse, so that the room creation is at the beginning of
+// the list, rather than the end.
+func (r *stateResolverV2) createPowerLevelMainline() []Event {
+	var mainline []Event
+
+	// Define our iterator function.
+	var iter func(event Event)
+	iter = func(event Event) {
+		// Append this event to the beginning of the mainline.
+		mainline = append([]Event{event}, mainline...)
```

Comment:
They're all slices. Slices have the underlying array, a size and a *capacity*, which doubles when things need to be resized via `append`.

That is to say, a naive insertion does not cause the entire array to be copied, and the idea of appending in a singular fashion like this is idiomatic Go code.

Reply:
This also has the peculiar side effect that the return arg from `append()` will *most of the time* be the same slice, but sometimes it'll be a nice slice if we had to resize.

Reply:
Right, but here it looks like it's creating a new array with the event, and appending the old array to that, which seems like it would do O(n) work for each iteration, as it needs to copy O(n) values over to the new array.